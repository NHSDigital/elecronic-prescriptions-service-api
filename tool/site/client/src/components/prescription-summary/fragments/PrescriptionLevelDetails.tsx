import {Bundle, CommunicationRequest, CommunicationRequestPayload, MedicationRequest} from "fhir/r4"
import {Field} from "formik"
import {Input, SummaryList} from "nhsuk-react-components"
import React from "react"
import {getPerformerSiteTypeExtension} from "../../../fhir/customExtensions"
import {getCurrentIssueNumberAndEndIssueNumber} from "../../../fhir/helpers"
import {COURSE_OF_THERAPY_TYPE_CODES, VALUE_SET_COURSE_OF_THERAPY_TYPE} from "../../../fhir/reference-data/valueSets"
import {formatCurrentDate, formatDate} from "../../../formatters/dates"
import {newLineFormatter} from "../../common/newLineFormatter"

function createPrescriptionLevelDetails(
  bundle: Bundle,
  medicationRequest: MedicationRequest
): PrescriptionLevelDetailsProps {
  const resources = bundle.entry.map(e => e.resource)
  const communicationRequests = resources.filter(r => r.resourceType === "CommunicationRequest") as Array<CommunicationRequest>

  const prescriptionId = medicationRequest.groupIdentifier.value

  const courseOfTherapyTypeCoding = VALUE_SET_COURSE_OF_THERAPY_TYPE.find(coding => coding.code === medicationRequest.courseOfTherapyType.coding[0].code)

  const authoredOn = formatCurrentDate()
  const startDate = formatDate(medicationRequest.dispenseRequest.validityPeriod?.start) ?? authoredOn
  const nominatedOds = medicationRequest.dispenseRequest?.performer?.identifier?.value || ""

  const nominatedTypeExtension = getPerformerSiteTypeExtension(medicationRequest.dispenseRequest.extension)
  const nominatedTypeCode = nominatedTypeExtension.valueCoding.code
  const nominatedType = getPharmacyTypeText(nominatedTypeCode)

  const patientInstructions = communicationRequests
    .flatMap(communicationRequest => communicationRequest.payload)
    .filter(Boolean)
    .filter(isContentStringPayload)
    .map(payload => payload.contentString)

  const detailsProps: PrescriptionLevelDetailsProps = {
    prescriptionId,
    courseOfTherapyType: courseOfTherapyTypeCoding.display,
    prescriptionTypeCode:
      medicationRequest.extension?.find(
        e => e.url === "https://fhir.nhs.uk/StructureDefinition/Extension-DM-PrescriptionType"
      )?.valueCoding.code,
    authoredOn,
    startDate,
    nominatedOds,
    nominatedType,
    patientInstructions
  }

  if (courseOfTherapyTypeCoding.code !== COURSE_OF_THERAPY_TYPE_CODES.ACUTE) {
    const [currentIssueNumber, endIssueNumber] = getCurrentIssueNumberAndEndIssueNumber(medicationRequest)
    detailsProps.currentIssueNumber = currentIssueNumber
    detailsProps.endIssueNumber = endIssueNumber
  }

  return detailsProps
}

interface PrescriptionLevelDetailsProps {
  prescriptionId: string
  courseOfTherapyType: string
  prescriptionTypeCode: string
  currentIssueNumber?: number
  endIssueNumber?: number
  authoredOn: string
  startDate: string
  nominatedOds?: string
  nominatedType?: string
  patientInstructions?: Array<string>
  editMode?: boolean
}

const SummaryListRow = ({label, value}: { label: string, value: string | JSX.Element | JSX.Element[] }) => {
  return (
    <SummaryList.Row>
      <SummaryList.Key>{label}</SummaryList.Key>
      <SummaryList.Value>{value}</SummaryList.Value>
    </SummaryList.Row>
  )
}

const PrescriptionLevelDetails = ({
  prescriptionId,
  courseOfTherapyType,
  prescriptionTypeCode,
  currentIssueNumber,
  endIssueNumber,
  authoredOn,
  startDate,
  nominatedOds,
  nominatedType,
  patientInstructions,
  editMode
}: PrescriptionLevelDetailsProps) => {
  const patientInstruction = newLineFormatter(patientInstructions)

  return (
    <SummaryList>
      <SummaryListRow label="ID" value={prescriptionId} />
      <SummaryListRow label="Prescription Type Code" value={courseOfTherapyType} />
      <SummaryListRow label="Course Of Therapy" value={prescriptionTypeCode} />

      {currentIssueNumber &&
        <SummaryListRow label="Issue Number" value={`${currentIssueNumber} of ${endIssueNumber}`} />
      }

      <SummaryListRow label="Authored On" value={authoredOn} />
      <SummaryListRow label="Effective Date" value={startDate} />

      {nominatedOds &&
        <>
          <SummaryListRow label="Nominated Pharmacy ODS Code" value={
            editMode
              ? <Field id="nominatedOds" name="nominatedOds" as={Input} width={30} />
              : nominatedOds
          } />

          <SummaryListRow label="Nominated Pharmacy Type" value={nominatedType} />
        </>
      }

      {patientInstructions.length > 0 &&
        <SummaryListRow label="Patient Instructions" value={patientInstruction} />
      }
    </SummaryList>
  )
}

function getPharmacyTypeText(code: string): string {
  console.log("performer code: " + code)
  const trimmedCode = code.trim()
  if (trimmedCode === "P1") {
    return "Other (e.g. Community Pharmacy)"
  } else if (trimmedCode === "P2") {
    return "Appliance Contractor"
  } else if (trimmedCode === "P3") {
    return "Dispensing Doctor"
  } else if (trimmedCode === "0004" || trimmedCode === "4" || trimmedCode === "foo") {
    return "None"
  } else {
    return "--- " + trimmedCode
  }
}

function isContentStringPayload(payload: CommunicationRequestPayload): boolean {
  return !!payload.contentString
}

export {
  PrescriptionLevelDetails,
  PrescriptionLevelDetailsProps,
  createPrescriptionLevelDetails
}
