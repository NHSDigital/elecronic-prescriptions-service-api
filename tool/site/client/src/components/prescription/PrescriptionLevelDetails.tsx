import {Input, SummaryList} from "nhsuk-react-components"
import React, {FC} from "react"
import {CommunicationRequest, CommunicationRequestPayload, MedicationRequest} from "fhir/r4"
import {formatCurrentDate, formatDate} from "../../formatters/dates"
import {getPerformerSiteTypeExtension} from "../../fhir/customExtensions"
import {newLineFormatter} from "../prescription-summary/newLineFormatter"
import {COURSE_OF_THERAPY_TYPE_CODES, VALUE_SET_COURSE_OF_THERAPY_TYPE} from "../../fhir/reference-data/valueSets"
import {getCurrentIssueNumberAndEndIssueNumber} from "../../fhir/helpers"
import {Field} from "formik"

function createPrescriptionLevelDetails(
  editMode: boolean,
  medicationRequest: MedicationRequest,
  communicationRequests?: Array<CommunicationRequest>
): PrescriptionLevelDetailsProps {
  const prescriptionId = medicationRequest.groupIdentifier.value

  const courseOfTherapyTypeCoding = VALUE_SET_COURSE_OF_THERAPY_TYPE.find(coding => coding.code === medicationRequest.courseOfTherapyType.coding[0].code)

  const authoredOn = formatCurrentDate()
  const startDate = formatDate(medicationRequest.dispenseRequest.validityPeriod?.start) ?? authoredOn
  const nominatedOds = medicationRequest.dispenseRequest?.performer?.identifier?.value || ""

  const nominatedTypeExtension = getPerformerSiteTypeExtension(medicationRequest.dispenseRequest.extension)
  const nominatedTypeCode = nominatedTypeExtension.valueCoding.code
  const nominatedType = getPharmacyTypeText(nominatedTypeCode)

  const patientInstructions = communicationRequests
    .flatMap(communicationRequest => communicationRequest.payload)
    .filter(Boolean)
    .filter(isContentStringPayload)
    .map(payload => payload.contentString)

  const detailsProps: PrescriptionLevelDetailsProps = {
    prescriptionId,
    courseOfTherapyType: courseOfTherapyTypeCoding.display,
    prescriptionTypeCode:
      medicationRequest.extension?.find(
        e => e.url === "https://fhir.nhs.uk/StructureDefinition/Extension-DM-PrescriptionType"
      )?.valueCoding.code,
    authoredOn,
    startDate,
    nominatedOds,
    nominatedType,
    patientInstructions,
    editMode
  }

  if (courseOfTherapyTypeCoding.code !== COURSE_OF_THERAPY_TYPE_CODES.ACUTE) {
    const [currentIssueNumber, endIssueNumber] = getCurrentIssueNumberAndEndIssueNumber(medicationRequest)
    detailsProps.currentIssueNumber = currentIssueNumber
    detailsProps.endIssueNumber = endIssueNumber
  }

  return detailsProps
}

export interface PrescriptionLevelDetailsProps {
  prescriptionId: string
  courseOfTherapyType: string
  prescriptionTypeCode: string
  currentIssueNumber?: number
  endIssueNumber?: number
  authoredOn: string
  startDate: string
  nominatedOds?: string
  nominatedType?: string
  patientInstructions?: Array<string>
  editMode: boolean
}

const PrescriptionLevelDetails: FC<PrescriptionLevelDetailsProps> = ({
  prescriptionId,
  courseOfTherapyType,
  prescriptionTypeCode,
  currentIssueNumber,
  endIssueNumber,
  authoredOn,
  startDate,
  nominatedOds,
  nominatedType,
  patientInstructions,
  editMode
}) => {
  const patientInstruction = newLineFormatter(patientInstructions)
  return (
    <SummaryList>
      <SummaryList.Row>
        <SummaryList.Key>ID</SummaryList.Key>
        <SummaryList.Value>{prescriptionId}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Course Of Therapy</SummaryList.Key>
        <SummaryList.Value>{courseOfTherapyType}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Prescription Type Code</SummaryList.Key>
        <SummaryList.Value>{prescriptionTypeCode}</SummaryList.Value>
      </SummaryList.Row>
      {currentIssueNumber &&
        <SummaryList.Row>
          <SummaryList.Key>Issue Number</SummaryList.Key>
          <SummaryList.Value>{currentIssueNumber} of {endIssueNumber}</SummaryList.Value>
        </SummaryList.Row>
      }
      <SummaryList.Row>
        <SummaryList.Key>Authored On</SummaryList.Key>
        <SummaryList.Value>{authoredOn}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Effective Date</SummaryList.Key>
        <SummaryList.Value>{startDate}</SummaryList.Value>
      </SummaryList.Row>
      {nominatedOds &&
        <>
          <SummaryList.Row>
            <SummaryList.Key>Nominated Pharmacy ODS Code</SummaryList.Key>
            <SummaryList.Value>
              {editMode
                ? <Field
                  id="nominatedOds"
                  name="nominatedOds"
                  as={Input}
                  width={30}
                />
                : nominatedOds
              }
            </SummaryList.Value>
          </SummaryList.Row>
          <SummaryList.Row>
            <SummaryList.Key>Nominated Pharmacy Type</SummaryList.Key>
            <SummaryList.Value>{nominatedType}</SummaryList.Value>
          </SummaryList.Row>
        </>
      }
      {patientInstructions.length > 0 &&
        <SummaryList.Row>
          <SummaryList.Key>Patient Instructions</SummaryList.Key>
          <SummaryList.Value>{patientInstruction}</SummaryList.Value>
        </SummaryList.Row>
      }
    </SummaryList>
  )
}

function getPharmacyTypeText(code: string): string {
  if (code === "P1") {
    return "Other (e.g. Community Pharmacy)"
  } else if (code === "P2") {
    return "Appliance Contractor"
  } else if (code === "P3") {
    return "Dispensing Doctor"
  } else if (code === "0004") {
    return "None"
  } else {
    return ""
  }
}

function isContentStringPayload(payload: CommunicationRequestPayload): boolean {
  return !!payload.contentString
}

export {
    PrescriptionLevelDetails,
    createPrescriptionLevelDetails,
}
