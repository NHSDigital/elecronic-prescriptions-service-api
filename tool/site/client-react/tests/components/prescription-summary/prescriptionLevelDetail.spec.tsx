import * as React from "react"
import {render} from "@testing-library/react"
import pretty from "pretty"

import {prescriptionLevelDetailProps} from "./props"
import PrescriptionLevelDetails from "../../../src/components/prescription-summary/prescriptionLevelDetails";

test("Renders correctly", async () => {
  const component = <PrescriptionLevelDetails {...prescriptionLevelDetailProps}/>
  const {container} = render(component)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
