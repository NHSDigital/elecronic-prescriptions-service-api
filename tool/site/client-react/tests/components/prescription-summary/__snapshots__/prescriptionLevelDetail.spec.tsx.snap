// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
"<dl class=\\"nhsuk-summary-list\\">
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">ID</dt>
    <dd class=\\"nhsuk-summary-list__value\\">A0548B-A99968-451485</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Repeat Information</dt>
    <dd class=\\"nhsuk-summary-list__value\\">1/6</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Authored on</dt>
    <dd class=\\"nhsuk-summary-list__value\\">01-11-2021</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Effective date</dt>
    <dd class=\\"nhsuk-summary-list__value\\">02-11-2021</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Nominated Pharmacy ODS Code</dt>
    <dd class=\\"nhsuk-summary-list__value\\">VNCEL</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Nominated Pharmacy Type</dt>
    <dd class=\\"nhsuk-summary-list__value\\">Other (e.g. Community Pharmacy)</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Patient Instructions</dt>
    <dd class=\\"nhsuk-summary-list__value\\">Take the medicine.</dd>
  </div>
</dl>"
`;
